### DISCUSSION TITLE
Change password by entering the old password

### DISCUSSION THREAD

**Discussion by Hungermatt** [supabase/supabase#11412] [Questions] [answered]
Hi,

I'm using supabase with flutter and I'm at a standstill with the "change password".
The problem is simply: when I want to change password, I want the app check the old password before entering the new one.
On 'auth/users' table there is the password encrypted, but it's useless.
How do I do?

Thanks for any suggestion

**Comment by GaryAustin1** [comment]
This thread might be of help with a function to verify and change.  https://github.com/supabase/supabase/discussions/4042
I'm not very familiar with the flutter client, but normally updateUser on the js client,needs to be recently signed in to change the password.

**Reply by Hungermatt** [reply]
Thanks for the reply. That is interesting: the problem I have posed myself is that, in the event of theft of the phone, it would still have the possibility to change the password without knowing the old one. So I thought of putting an extra security factor so that the password is not changed easily.

I'll try this solution and let you know. Just in case, I'll put the piece of code in flutter that takes advantage of that function you posted.

**Comment by faizank09855** [comment]
```
you can Enter password on TextField  , 
after That you have to execute login Auth Function Once Again  , 
if login Successfully you can change the new password .
```

**Comment by Hungermatt** [comment, marked as answer]
### Success!
After many tries (lasting two days), I found the correct way to do it. This is valid for every platform: I created a function on supabase done in this way

> ```sql
> create or replace function changepassword("current_plain_password" text, "new_plain_password" text, "current_id" uuid)
> returns varchar
> language plpgsql
> security definer
> as $$
> DECLARE
> encpass auth.users.encrypted_password%type;
> BEGIN
>   SELECT encrypted_password
>   FROM auth.users
>   INTO encpass
>   WHERE id = current_id and encrypted_password = crypt(current_plain_password, auth.users.encrypted_password);
>
>   -- Check the currect password and update
>   IF NOT FOUND THEN
>     return 'incorrect';
>   else
>     UPDATE auth.users SET encrypted_password = crypt(new_plain_password, gen_salt('bf')) WHERE id = current_id;
>     return 'success';
>   END IF;
>
> END;
> $$
> ```

###  How to use?

On personal code of the app, just recall the function on a variable like this way (this is for flutter app):

> ```dart
> final data = await supabase.rpc('changepassword', params: {'current_plain_password': OldPass, 'new_plain_password': NewPass, 'current_id': IdSessione});

where;
- _current_plain_password_ is the old password
- _new_plain_password_ is the new password
- _current_id_ is the id in the current session of the app.

That function returns _incorrect_ if the password is wrong, _success_ if the password is correct and the function updates correctly the password.

NB: _encpass_ variable is only needed to avoid making the function go wrong and guarantee its 100% success. Without it, I don't know if it will work (I have made several attempts and this works correctly)

I hope it can help others too

**Reply by Hungermatt** [reply]
Thanks @GaryAustin1 for your suggestion ;)

**Reply by denvudd** [reply]
Hey @Hungermatt 
Seems like a nice solution to me, did it prove itself over time?

**Comment by zivtamary** [comment]
My solution was to create another table called User, when a user signs up I add him to this table with his password (hashed)
I then created a server route for password update, in there I verify the hashed current password against the password he entered, if it matches I continue to use the supabase update user method, if not I return 400 bad request.
